Sign up 페이지 완료

CSE-pebble opened the pull request:
2024.02.03
Sign up 페이지 완료

### Issue

**1. 코드 리팩토링을 했으나 여전히 지저분한 index.jsx 파일.............**
- 특히 Input 컴포넌트를 이렇게 빼는게 맞는지..? 

**2. handleValidation, handleErrorMsg가 뭔가 최선의 코드가 아닌 것 같음.......** 

**3. 함수 하나의 길이가 너무 길다.** 

**4.  아이디, 닉네임, 비밀번호, 핸드폰 번호 형식**
- 회의를 통해 확실히 정해야할 것 같다!
- 특히 최소, 최대 길이 제한이 필요할 듯

**5. CSS) Header의 "회원가입" 글자가 X 버튼 때문에 가운데로 안옴..**

---
⭐ Sign up 페이지 생성
* Sign up 페이지 생성 후 NavBar에 연결
<br>

⭐ 기능
* X 버튼 클릭 시 이전 페이지로 이동
* '비밀번호'와 '비밀번호 확인'에 입력한 값이 다를 시 오류 메세지 생성
* '핸드폰 번호'에 숫자 외의 문자 입력 시 오류 메세지 생성  
<br>

⭐ 기타
* 필요없는 CSS 일부 삭제 (App.css, index.css)

CSE-pebble committed:
Env: Inital Branch Setting
- SignUp 브랜치 생성
- SingUp 디렉토리 생성

CSE-pebble committed:
Design: Add SignUp Screen
- 회원가입 화면 퍼블리싱

CSE-pebble committed:
Feat: Connect Link in NavBar(signUp)

CSE-pebble committed:
Feat: Connect Link in NavBar(signUp)

CSE-pebble committed:
Feat: Add CloseBtn Navigate
- 닫기 버튼을 누르면 이전 페이지로 이동

CSE-pebble committed:
Design: Remove Vite initial CSS
- Vite에서 기본적으로 제공된 CSS 지우기

CSE-pebble committed:
Fix: Change Button Name
- Button -> DuplicateCheckButton으로 버튼명 변경

CSE-pebble committed:
Feat: Add Phone number type Check
- 핸드폰 번호 input에 숫자 외 입력하면 에러 메세지 보임

CSE-pebble committed:
Feat: Add Password Confirm check
- 비밀번호 input, 비밀번호 확인 input 값이 일치하지 않으면 에러 메세지 생성

CSE-pebble committed:
Fix: Add phoneNumber value to store
- 핸드폰 번호를 저장하여 넘기기 위한 phoneNumber 변수 생성

CSE-pebble committed:
Feat: Add Profile Image Upload function & limitation
- 프로필 이미지를 클릭하면 원하는 이미지를 업로드 할 수 있음
- 이미지 제한 : 최대 5MB, 형식(jpg/jepg/png)

CSE-pebble committed:
Feat: Add nickname duplicate check
- 닉네임에 공백이 포함될 시 중복 확인 버튼 비활성화, 에러 메세지 출력
- 닉네임을 입력하지 않고 중복 확인 버튼을 누를 시 에러 메세지 출력
- 중복 닉네임을 입력하고 중복 확인 버튼을 누를 시 에러 메세지 출력
- 중복 없는 닉네임을 입력하고 중복 확인 버튼을 누를 시 통과 메세지 출력

CSE-pebble committed:
Fix: Modify function name
- 오타 SingUp->SignUp

CSE-pebble committed:
Feat: Add id duplicate check
- 아이디에 숫자/영어 외 문자 or 공백이 포함될 시 중복 확인 버튼 비활성화, 에러 메세지 출력
- 아이디를 입력하지 않고 중복 확인 버튼을 누를 시 에러 메세지 출력
- 중복 아이디를 입력하고 중복 확인 버튼을 누를 시 에러 메세지 출력
- 중복 없는 아이디를 입력하고 중복 확인 버튼을 누를 시 통과 메세지 출력

CSE-pebble committed:
Fix: Fix nickname/id check error
- 새로 입력하면 무조건 에러 메세지와 중복확인 체크 변수가 초기화 되도록 수정

CSE-pebble committed:
Fix: Modify value name
- 컨벤션에 맞게 setNicknameChecked 변수명을 setIsNicknameChecked로 수정

CSE-pebble committed:
Feat: Add SignUpButton style

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -0,0 +1,439 @@
+        "영어와 숫자만 입력 가능하며, 공백을 포함할 수 없습니다."
+      );
+      setIdButtonDisabled(true);
+      setIsNicknameChecked(false);
```
아이디를 조건에 맞지 않게 입력했을 경우 닉네임 중복 여부를 다시 체크 해야한다고 생각하신 이유가 뭔지 궁금합니다!

CSE-pebble replied:
> 아이디를 조건에 맞지 않게 입력했을 경우 닉네임 중복 여부를 다시 체크 해야한다고 생각하신 이유가 뭔지 궁금합니다!

헉 복사하고 실수로 수정을 안한 것 같아요..! 닉네임이 아니라 아이디 입니당..

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -0,0 +1,439 @@
+            {nicknameErrorMessage && (
+              <ErrorMsg>{nicknameErrorMessage}</ErrorMsg>
+            )}
+            {isNicknameChecked && <PassMsg>사용 가능한 닉네임입니다.</PassMsg>}
```
사용자를 고려한 에러 메시지, 안내 메시지 너무 좋아요!

YeonjuSeo commented on package.json:
```diff
@@ -12,6 +12,7 @@
   "dependencies": {
     "react": "^18.2.0",
     "react-dom": "^18.2.0",
+    "react-icons": "^5.0.1",
```
사용하는 곳이 없는 패키지라면 삭제 부탁드립니다!

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -0,0 +1,439 @@
+  overflow-y: auto;
+`;
+
+const Container = styled.div`
```
Wrapper와 의미 구별이 좀 더 잘 되는 이름으로 지어주시면 더 좋을 것 같습니다!

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -0,0 +1,439 @@
+              onChange={handleImageChange}
+            />
+          </Profile>
+        </ProfileWrapper>
```
이미지 프로필도 회원가입 시 함께 백엔드로 제출 되어야 하므로
이 요소 또한 Form 요소 안에 있는 것이 적절하다고 생각되는데요, 혹시 이렇게 설정하신 이유가 있으실까요?

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -0,0 +1,439 @@
+  const [idErrorMessage, setIdErrorMessage] = useState("");
+  const [idButtonDisabled, setIdButtonDisabled] = useState(false);
+  // 가입 완료하기
+  const [signUpButtonDisabled, setSignUpButtonDisabled] = useState(true);
```
수많은 state들은 필요한 요소이긴 하지만 다소 많게 느껴지는 것도 사실인 것 같습니다.
이 부분은 [이 글](https://velog.io/@yujng/React-useState-이용해서-회원가입-구현하기) 을 참고하시면 개선하는 데에 도움이 되지 않을까 싶습니다.

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -0,0 +1,439 @@
+  // 비밀번호
+  const [password, setPassword] = useState("");
+  const [passwordConfirm, setPasswordConfirm] = useState(""); //실제 넘길 값
+  const [passwordErrorMsg, setPasswordErrorMsg] = useState("");
```
아래의 코드를 쭉 읽어 보니 에러 메시지들에 들어가야 할 내용은 정해진 것으로 해석됩니다.
저라면 예상치 못한 변경 사항이 발생하기 쉬운 문자열보다는 true, false 값을 받는 state로 설정하고, 그 값에 따라 정해진 메시지를 출력하도록 구현했을 것 같습니다! 

나현 님께서 205번째 줄에서 하셨던 것처럼
`{isNicknameChecked && <PassMsg>사용 가능한 닉네임입니다.</PassMsg>}` 이렇게요!

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -0,0 +1,439 @@
+      setNicknameErrorMessage("닉네임을 입력해주세요.");
+      setIsNicknameChecked(false);
+      return;
+    }
```
꼼꼼하게 함수를 잘 짜셨지만, 닉네임과 아이디의 공백 처리 등 유사한 함수가 반복되는 것도 보입니다.
유사한 함수를 줄여 최대한 한 파일 내의 코드 수가 길어지지 않도록 유의해주세요!

유명한 책 '클린 코드'에 따르면 한 함수 내부 코드 길이로 권장되는 길이는 3줄입니다!
현실적으로 모든 함수를 3줄로 짜는 것은 어려울 뿐더러, 그것이 무조건 좋다고만 할 수는 없습니다. 하지만 그정도로 코드가 길어지는 것은 경계할 만한 사항이라고 알아주시면 좋겠습니다!

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -0,0 +1,439 @@
+          <InputWrapper>
+            <InputTitle>
+              닉네임<span>*</span>
+            </InputTitle>
```
return 내부의 코드가 긴 편에 속합니다. 이런 경우 가독성을 해쳐 정작 화면의 중요한 부분을 확인하기가 어려워지기 쉽습니다.

이렇게 매 항목마다 반복되는 컴포넌트는 가독성 증가를 위해 동일한 폴더 내 컴포넌트로 분리하신 뒤 임포트해서 사용하시면 가독성을 높이는 데에 기여할 수 있을 것입니다!

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -0,0 +1,439 @@
+                disabled={nicknameButtonDisabled}
+              >
+                중복 확인
+              </DuplicateCheckButton>
```
공통된 styled component를 활용하신 것이 좋았습니다!
그런 만큼 이것 또한 동일 디렉터리 내 컴포넌트 파일로 분리하여 임포트 해와서 쓰면 어떨지도 고려해주시면 좋겠습니다!

CSE-pebble committed:
Fix: Remove react-icon package
- 사용하지 않는 패키지 삭제

CSE-pebble committed:
Fix: Move ProfileWrapper in Form
- ProfileWrapper를 Form 태그 안으로 옮김

CSE-pebble committed:
Fix: Fix typo

CSE-pebble committed:
Refactor: Separate Header.jsx component
- Header 컴포넌트를 jsx 파일로 분리

CSE-pebble committed:
Refactor: Separate Profile.jsx component
- Profile 컴포넌트를 jsx 파일로 분리

CSE-pebble committed:
Refactor: Separate DuplicateCheckButton.jsx
- DuplicateCheckButton 컴포넌트를 jsx 파일로 분리

CSE-pebble committed:
[Ing] Refactor: Separate Input component
- Input 컴포넌트를 jsx 파일로 분리

CSE-pebble committed:
Refactor: Separate Input component
- Input 컴포넌트를 jsx 파일로 분리(완료)

CSE-pebble committed:
Refactor: Separate SignUpButton component
- SignUpButton 컴포넌트를 jsx 파일로 분리

CSE-pebble committed:
Refactor: Add form, valid state & Refactor nickname code
- input 관련 state들을 객체로 한꺼번에 관리하는 form state 생성
- input의 유효성을 객체로 한꺼번에 관리하는 valid state 생성 (이후 수정할 수도 있음)
- 새로운 state 생성에 따라 nickname 관련 코드 수정

CSE-pebble committed:
Refactor: Refactor id code
- 새로운 state 생성에 따라 id 관련 코드 수정

CSE-pebble committed:
Refactor: Refactor password code
- 새로운 state 생성에 따라 password 코드 수정

CSE-pebble committed:
Refactor: Refactor phoneNumber code
- 새로운 state 생성에 따라 phoneNumber 코드 수정

CSE-pebble committed:
Feat: Add value property to password/passwordCofirm
- 비밀번호 input란 초기화를 위한 value 속성 추가

CSE-pebble committed:
Refactor: Refactor validation reset code
- input란에 새로 내용 입력하면 해당 valid 값 초기화

CSE-pebble committed:
Feat: Add phoneNumber valid check
- 핸드폰 번호 valid 값 업데이트

CSE-pebble committed:
Refactor: Add errorMsg state
- 에러메세지 관련 state들을 객체로 한꺼번에 관리하는 errorMsg state 생성
- 그에 따른 관련 코드 리팩토링

CSE-pebble committed:
Fix: Fix password check
- 비밀번호란을 이후에 수정할 경우 에러 메세지가 뜨지 않는 버그 수정

CSE-pebble committed:
Style: Change value name

CSE-pebble committed:
Feat: Check form for SignUpButton Enabled
- form의 모든 key의 값이 존재하면 가입 완료하기 버튼 활성화

CSE-pebble committed:
Feat: Valid check if SignUpButton is clicked
- 가입 완료하기 버튼을 누르면 유효성 검사

CSE-pebble committed:
Feat: Add default profile image setting

CSE-pebble committed:
Style: Remove unused log
- 쓸모없는 로그 삭제

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -1,266 +1,312 @@
-        } else {
-          alert(
-            "지원되지 않는 형식입니다. jpg, jpeg, png 이미지 형식을 사용해주세요."
-          );
-        }
-      }
-    }
+  const handleErrorMsg = (key, value) => {
+    setErrorMsg((prevError) => ({ ...prevError, [key]: value }));
```
더 나은 코드를 고민하시는 게 너무 멋지십니다!!🥺
Issue에는 최선의 코드가 아닌 것 같다고 하셨는데 왜 그렇게 생각하셨는지가 궁금해지네요! 제가 생각하지 못하는 방향의 개선점을 생각하셨을 수도 있을 것 같다는 생각이 들어서요!

혹시 
1. 두 개의 로직이 비슷해보여서 더 일반화 할 수 있지 않을까 하는 생각이 드셨다면? 저는 두 개는 하는 일이 다르기 때문에 이 경우에는 일반화를 하지 않는 것이 더 좋다고 생각했습니다!
2. 함수 내부가 코드 한 줄이라(함수가 하는 일이 너무 단순) 고민이 드셨다면? 저는 가독성을 위해서라도 충분히 좋은 선택이라고 생각합니다!

다르게 생각하셨다면 언제든 말씀해주세요. 나현 님의 생각이 궁금해요!

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -1,266 +1,312 @@
-  // 프로필 사진
-  const [profileImage, setProfileImage] = useState(null); // 실제 넘길 값
+  // 넘길 값
+  const [form, setForm] = useState({
+    profile: null,
+    nickname: "",
+    id: "",
+    password: "",
+    passwordConfirm: "",
+    phoneNumber: "",
+  });
+  // 유효성 검사
+  const [valid, setValid] = useState({
+    nickname: false,
+    id: false,
+    password: false,
+    phoneNumber: false,
+  });
+  // 에러 메세지
+  const [errorMsg, setErrorMsg] = useState({
+    nickname: "",
+    id: "",
+    password: "",
+    phoneNumber: "",
+  });
```
항목별로 state를 분류해주신 덕분에 가독성이 엄청 좋아졌어요!!!

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -1,266 +1,312 @@
-      setNicknameButtonDisabled(true); // 공백이 있으면 버튼 비활성화
+      handleErrorMsg("nickname", "닉네임에 공백을 포함할 수 없습니다.");
+      handleValidation("nickname", false);
+      setIsNicknameBtnDisabled(true); // 공백이 있으면 버튼 비활성화
```
state 값을 단순히 toggle 하지 않고 이렇게 직접 넣는 형태로 구현해주셨는데요, 저는 이 방향이 의도치 않은 오류를 방지하기에도 무척 좋은 방향이라고 생각합니다 너무 좋아요!

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -1,266 +1,312 @@
-      setNicknameErrorMessage(""); // 에러 메시지 초기화
-      setIsNicknameChecked(true); // PassMsg를 표시하기 위해 상태 업데이트
+      handleErrorMsg("nickname", ""); // 에러 메시지 초기화
+      handleValidation("nickname", true); // PassMsg를 표시하기 위해 상태 업데이트
     } else {
-      setNicknameErrorMessage("이미 사용 중인 닉네임입니다.");
-      setIsNicknameChecked(false);
+      handleErrorMsg("nickname", "이미 사용 중인 닉네임입니다.");
+      handleValidation("nickname", false);
     }
```
이러한 형태는 삼항 연산자를 사용하는 방식도 고려해보실 수 있을 것 같습니다!
예를 들면
```
handleErrorMsg("nickname", isNicknameAvailable === true ? "" : "이미 사용 중인 닉네임입니다.");
handleValidation("nickname", isNicknameAvailable);
```

코드가 짧아진다고 무조건 다 좋은 것은 아니지만 사람에 따라 이러한 형태가 더 가독성이 좋다고 느끼실 수도 있을 것 같아서 제안드렸습니다.

또한 다양한 경우에 에러 메시지를 텍스트로 활용하고 계신데요, 필수적인 것은 아니지만 이러한 경우 사용하기 좋은 것이 enum 타입일 것 같습니다. 혹시 궁금하시다면 [이 글](https://gobae.tistory.com/96)이 참조하시기 좋을 듯 합니다. enum 타입은 여러 군데에서 사용하거나 너무 내용이 긴 경우 따로 파일을 만들어서 사용하기도 합니다!

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -1,266 +1,312 @@
-                onChange={handleIdChange}
-              />
+                disabled={isNicknameBtnDisabled}
+              ></DuplicateCheckButton>
```
이런 경우는 꼭 태그를 직접 닫으실 필요 없이 셀프 클로징 태그로 활용하시면 됩니다!
예를 들면 `<DuplicatedCheckButton onClick={} disabled={} />` 이렇게요!

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -1,266 +1,312 @@
+              ></DuplicateCheckButton>
+            }
+            errorMsg={errorMsg.id}
+            passMsg={valid.id && "사용 가능한 아이디입니다."}
```
이런 문구도 위에서 말씀드렸던 enum 타입을 통해 정리하실 수 있습니다. 예를 들어

```
// 선언 부분
const passMsg = {
ID: "사용 가능한 아이디입니다.",
NICKNAME:"사용 가능한 닉네임입니다.:
};
// 실제 코드 적용 부분
...
passMsg={passMsg.NICKNAME}
```

YeonjuSeo commented on src/pages/SignUp/Input.jsx:
```diff
@@ -0,0 +1,62 @@
+        <span>*</span>
+      </Title>
+      <ContentsWrapper>
+        <Text {...text}></Text>
```
Text라는 이름은 선언 부분을 봐서는 input 태그의 스타일링명이라고 연상하기가 쉽지만은 않은 부분이 있습니다. 최소한 input이라는 특성이 드러나는 이름이면 더 좋겠습니다!

YeonjuSeo commented on src/pages/SignUp/Input.jsx:
```diff
@@ -0,0 +1,62 @@
+import styled from "styled-components";
+
+const Input = ({ title, text, duplicate, errorMsg, passMsg }) => {
```
이 컴포넌트는 단순 input 태그 외에도 포함하는 내용이 많아 보입니다(제목, 입력창, 중복 메시지, 통과 메시지 등...) 저라면 순수하게 스타일링 된 input 태그에 관한 컴포넌트에만 Input이라는 이름을 붙이고 이 컴포넌트에는 또다른 이름을 고민해볼 것 같습니다.

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -1,266 +1,312 @@
+
+          <Input
+            title="아이디"
+            text={{
+              type: "text",
+              placeholder: "아이디를 입력해주세요",
+              name: "id",
+              onChange: (e) => {
+                handleChange(e);
+                handleIdChange(e);
+              },
+            }}
```
태그의 세부 어트리뷰트를 넘기는 부분에서 고민이 많으셨을 것 같습니다. 저도 더 나은 방향이 뭘지 고민이 되네요. 제가 고민이 되는 이유는 
1. 책 '클린코드'에 따르면 함수의 프로퍼티는 2개 이하를 지향하기 때문입니다. 
2. 그러기 위해 컴포넌트를 너무 잘게 쪼갤 경우 같은 프로퍼티를 거듭 부모에서 자식으로 전달하는 Prop drilling이 발생하기 쉽기 때문입니다.
3. 대개 Prop Drilling을 해결하기 위해서는 상태관리 툴을 사용하는데 이 부분은 단순 컴포넌트를 구성하기 위한 값들이라서 상태 관리 툴을 사용할 부분이라는 생각은 잘 들지 않기 때문입니다.

너무 복잡하죠??ㅋㅋㅋㅋㅋ 저도 정말 고민이 되네요.

우선 저라면(지금 생각 나는 방법은)
1. 코드가 좀 길어져도 저는 text에 객체를 통으로 넘기고 
2. 대신 type의 경우 일반적으로 text인 경우가 많으므로 컴포넌트 내 default 값으로 text를 설정해두고 예외적인 경우에만 입력을 받을 것 같습니다.(이런 식으로 최대한 컴포넌트에 입력되는 프로퍼티 값을 줄이는 방향을 구상할 것 같습니다.)

더 좋은 방향이 떠오른다면 말씀해주세요!!

이렇게 해서 약간의 Prop drilling이 발생하더라도 값을 넘기고,

YeonjuSeo commented on src/pages/SignUp/Header.jsx:
```diff
@@ -0,0 +1,37 @@
+}
+
+const Wrapper = styled.header`
+  display: flex;
+  align-items: center;
```
피그마와 같이 왼쪽은 비어있고 중간에 제목, 우측에 버튼이 있는 배치를 하고 싶으시다면 `display:grid` 속성을 사용해보시는 건 어떨까요?

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -1,266 +1,312 @@
-
     const userInput = e.target.value;
-
     if (/[^a-zA-Z0-9]/.test(userInput) || userInput.includes(" ")) {
```
Issue에 써주신대로 최소, 최대 길이 제한 좋아요!
근데 특별히 제한이 필요하다고 생각하시게 된 계기가 있는지 궁금해서 회의 때 말씀해주시면 좋겠어요!!

YeonjuSeo commented on src/pages/SignUp/index.jsx:
```diff
@@ -1,266 +1,312 @@
+    if (!form.id) {
+      handleErrorMsg("id", "아이디를 입력해주세요.");
+      handleValidation("id", false);
       return;
```
early Return 방식은 함수 내 복잡도를 줄이고 이해를 더 쉽게 해서 적용해주신 거 너무 좋습니다!! LGTM☺️

YeonjuSeo reviewed (commented):
와 코드가 100줄이나 줄었어요!!! 함께 더 좋은 코드를 위해 고민해주셔서 너무너무 좋았습니다! 일부 함수에서는 비슷비슷한 로직이 보이는데, 이걸 어떻게 또 잘 일반화하고 추상화 하면 대폭 줄일 수 있을 것 같기도 했습니다. LGTM💚
